[Bug]: display-name false positive when React is shadowed

### Is there an existing issue for this?

- [x] I have searched the existing issues and my issue is unique
- [x] My issue appears in the command-line and not only in the text editor

### Description Overview

<!--Brief description-->

`react/display-name` will report an error `Component definition is missing display name` when the `React` global import is shadowed within a function scope.

The same error does not occur for shadowed `memo` or `forwardRef` though.

#### Example

```jsx
import React, { memo, forwardRef } from 'react'

const MixedShadowed = function () {
  const memo = (cb) => cb()
  const { forwardRef } = { forwardRef: () => null }
  const [React] = [{ memo, forwardRef }]

  const Comp = memo(() => {
    return <div>shadowed</div>
  })
  const ReactMemo = React.memo(() => null)
  const ReactForward = React.forwardRef((props, ref) => {
    return `${props} ${ref}`
  })
  const OtherComp = forwardRef((props, ref) => `${props} ${ref}`)

  return [Comp, ReactMemo, ReactForward, OtherComp]
}
```

Running eslint from the command line you'll see the error:

```
  11:21  error  Component definition is missing display name  react/display-name
  12:24  error  Component definition is missing display name  react/display-name
```

### Expected Behavior

I know most people won't shadow `React`, `memo` or `forwardRef`, but I'd still expect the error to not be reported in this case.


### eslint-plugin-react version

v7.37.5

### eslint version

v9.26.0

### node version

v22.15.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ This change log adheres to standards from [Keep a CHANGELOG](https://keepachange
 * Remove extra space from CLI warning ([#3942][] @junaidkbr)
 * [`jsx-key`]: detect missing keys in return statement with ternary operator ([#3928][] @hyeonbinHur)
 * [`jsx-key`]: detect missing keys in logical expressions ([#3986][] @yalperg)
+* [`display-name`]: avoid false positive when React is shadowed ([#3926][] @hyeonbinHur)
 
 ### Changed
 * [Docs] [`no-array-index-key`]: add template literal examples ([#3978][] @akahoshi1421)
@@ -36,6 +37,7 @@ This change log adheres to standards from [Keep a CHANGELOG](https://keepachange
 [#3942]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3942
 [#3930]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3930
 [#3928]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3928
+[#3926]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3926
 [#3923]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3923
 [#3877]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3877
 [#3729]: https://github.com/jsx-eslint/eslint-plugin-react/pull/3729

--- a/lib/rules/display-name.js
+++ b/lib/rules/display-name.js
@@ -161,6 +161,123 @@ module.exports = {
       return false;
     }
 
+    function hasVariableDeclaration(node, name) {
+      if (!node) return false;
+
+      if (node.type === 'VariableDeclaration') {
+        return node.declarations.some((decl) => {
+          if (!decl.id) return false;
+
+          // const name = ...
+          if (decl.id.type === 'Identifier' && decl.id.name === name) {
+            return true;
+          }
+
+          // const [name] = ...
+          if (decl.id.type === 'ArrayPattern') {
+            return decl.id.elements.some(
+              (el) => el && el.type === 'Identifier' && el.name === name
+            );
+          }
+
+          // const { name } = ...
+          if (decl.id.type === 'ObjectPattern') {
+            return decl.id.properties.some(
+              (prop) => prop.type === 'Property' && prop.key && prop.key.name === name
+            );
+          }
+
+          return false;
+        });
+      }
+
+      if (node.type === 'BlockStatement' && node.body) {
+        return node.body.some((stmt) => hasVariableDeclaration(stmt, name));
+      }
+
+      return false;
+    }
+
+    function isIdentifierShadowed(node, identifierName) {
+      let currentNode = node;
+
+      while (currentNode && currentNode.parent) {
+        currentNode = currentNode.parent;
+
+        if (
+          currentNode.type === 'FunctionDeclaration'
+          || currentNode.type === 'FunctionExpression'
+          || currentNode.type === 'ArrowFunctionExpression'
+        ) {
+          if (currentNode.body && hasVariableDeclaration(currentNode.body, identifierName)) {
+            return true;
+          }
+        }
+
+        if (currentNode.type === 'BlockStatement') {
+          if (hasVariableDeclaration(currentNode, identifierName)) {
+            return true;
+          }
+        }
+
+        if (
+          (currentNode.type === 'FunctionDeclaration'
+           || currentNode.type === 'FunctionExpression'
+           || currentNode.type === 'ArrowFunctionExpression')
+          && currentNode.params
+        ) {
+          const isParamShadowed = currentNode.params.some((param) => {
+            if (param.type === 'Identifier' && param.name === identifierName) {
+              return true;
+            }
+            if (param.type === 'ObjectPattern') {
+              return param.properties.some(
+                (prop) => prop.type === 'Property' && prop.key && prop.key.name === identifierName
+              );
+            }
+            if (param.type === 'ArrayPattern') {
+              return param.elements.some(
+                (el) => el && el.type === 'Identifier' && el.name === identifierName
+              );
+            }
+            return false;
+          });
+
+          if (isParamShadowed) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+    /**
+     * Checks whether the component wrapper (e.g. React.memo or forwardRef) is shadowed in the current scope.
+     * @param {ASTNode} node - The CallExpression AST node representing a potential component wrapper.
+     * @returns {boolean} True if the wrapper identifier (e.g. 'React', 'memo', 'forwardRef') is shadowed, false otherwise.
+     */
+    function isShadowedComponent(node) {
+      if (!node || node.type !== 'CallExpression') {
+        return false;
+      }
+
+      if (
+        node.callee.type === 'MemberExpression'
+        && node.callee.object.name === 'React'
+      ) {
+        return isIdentifierShadowed(node, 'React');
+      }
+
+      if (node.callee.type === 'Identifier') {
+        const name = node.callee.name;
+        if (name === 'memo' || name === 'forwardRef') {
+          return isIdentifierShadowed(node, name);
+        }
+      }
+
+      return false;
+    }
+
     // --------------------------------------------------------------------------
     // Public
     // --------------------------------------------------------------------------
@@ -269,9 +386,9 @@ module.exports = {
       'Program:exit'() {
         const list = components.list();
         // Report missing display name for all components
-        values(list).filter((component) => !component.hasDisplayName).forEach((component) => {
-          reportMissingDisplayName(component);
-        });
+        values(list)
+          .filter((component) => !isShadowedComponent(component.node) && !component.hasDisplayName)
+          .forEach((component) => { reportMissingDisplayName(component); });
         if (checkContextObjects) {
           // Report missing display name for all context objects
           forEach(

--- a/tests/lib/rules/display-name.js
+++ b/tests/lib/rules/display-name.js
@@ -31,6 +31,72 @@ ruleTester.run('display-name', rule, {
   valid: parsers.all([
     {
       code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const TestComponent = function () {
+          {
+            const memo = (cb) => cb()
+            const forwardRef = (cb) => cb()
+            const React = { memo, forwardRef }
+
+            const BlockMemo = memo(() => <div>shadowed</div>)
+            const BlockForwardRef = forwardRef(() => <div>shadowed</div>)
+            const BlockReactMemo = React.memo(() => <div>shadowed</div>)
+          }
+          return null
+        }
+      `,
+    },
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const Test1 = function (memo) {
+          return memo(() => <div>param shadowed</div>)
+        }
+
+        const Test2 = function ({ forwardRef }) {
+          return forwardRef(() => <div>destructured param</div>)
+        }
+      `,
+    },
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const TestComponent = function () {
+          function innerFunction() {
+            const memo = (cb) => cb()
+            const React = { forwardRef }
+
+            const Comp = memo(() => <div>nested</div>)
+            const ForwardComp = React.forwardRef(() => <div>nested</div>)
+            return [Comp, ForwardComp]
+          }
+          return innerFunction()
+        }
+      `,
+    },
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const MixedShadowed = function () {
+          const memo = (cb) => cb()
+          const { forwardRef } = { forwardRef: () => null }
+          const [React] = [{ memo, forwardRef }]
+
+          const Comp = memo(() => <div>shadowed</div>)
+          const ReactMemo = React.memo(() => null)
+          const ReactForward = React.forwardRef((props, ref) => \`\${props} \${ref}\`)
+          const OtherComp = forwardRef((props, ref) => \`\${props} \${ref}\`)
+
+          return [Comp, ReactMemo, ReactForward, OtherComp]
+        }
+      `,
+    },
+    {
+      code: `
         var Hello = createReactClass({
           displayName: 'Hello',
           render: function() {
@@ -848,6 +914,82 @@ ruleTester.run('display-name', rule, {
   ]),
 
   invalid: parsers.all([
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const TestComponent = function () {
+          {
+            const BlockReactMemo = React.memo(() => {
+              return <div>not shadowed</div>
+            })
+
+            const BlockMemo = memo(() => {
+              return <div>not shadowed</div>
+            })
+
+            const BlockForwardRef = forwardRef((props, ref) => {
+              return \`\${props} \${ref}\`
+            })
+          }
+
+          return null
+        }
+      `,
+      errors: [
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+      ],
+    },
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const Test1 = function () {
+          const Comp = memo(() => <div>not param shadowed</div>)
+          return Comp
+        }
+
+        const Test2 = function () {
+          function innerFunction() {
+            const Comp = memo(() => <div>nested not shadowed</div>)
+            const ForwardComp = React.forwardRef(() => <div>nested</div>)
+            return [Comp, ForwardComp]
+          }
+          return innerFunction()
+        }
+      `,
+      errors: [
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+      ],
+    },
+    {
+      code: `
+        import React, { memo, forwardRef } from 'react'
+
+        const MixedNotShadowed = function () {
+          const Comp = memo(() => {
+            return <div>not shadowed</div>
+          })
+          const ReactMemo = React.memo(() => null)
+          const ReactForward = React.forwardRef((props, ref) => {
+            return \`\${props} \${ref}\`
+          })
+          const OtherComp = forwardRef((props, ref) => \`\${props} \${ref}\`)
+
+          return [Comp, ReactMemo, ReactForward, OtherComp]
+        }
+      `,
+      errors: [
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+        { messageId: 'noDisplayName' },
+      ],
+    },
     {
       code: `
         var Hello = createReactClass({

